compiler/luci-interpreter/src/kernels/LQFullyConnected.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/LQFullyConnected.h"
+
+#include "kernels/Utils.h"
+
+#include <stdexcept>
+
+namespace luci_interpreter
+{
+namespace kernels
+{
+
+QuantizedBinaryInput::QuantizedBinaryInput(int32_t input_vec_size, const Tensor *input_scales)
+{
+  input_float_size = input_vec_size;
+  input_bin_size = ceil_div(input_float_size, 32);
+  encode_bits = input_scales->shape().dim(0);
+
+  // count of different levels is 2^(encode_bits)
+  quantized_levels = 1 << encode_bits;
+
+  // allocate buffers
+  input_binary = std::make_unique<int32_t[]>(input_bin_size * encode_bits);
+  quantization_thresholds = std::make_unique<float[]>(quantized_levels);
+
+  // init quantization levels
+  const float *scales = input_scales->data<float>();
+  for (int32_t i = 0; i < quantized_levels; ++i)
+  {
+    // TODO investigate faster variants to perform this
+    quantization_thresholds[i] = 0.0f;
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      quantization_thresholds[i] += (i | (1 << b)) ? scales[b] : -scales[b];
+    }
+  }
+
+  // compute quantization thresholds via levels
+  float last_v = quantization_thresholds[0];
+  quantization_thresholds[0] = std::numeric_limits<float>::min(); // -INF
+
+  for (uint32_t i = 1; i < quantized_levels; ++i)
+  {
+    float curr_v = quantization_thresholds[i];
+    quantization_thresholds[i] = (curr_v + last_v) / 2.0f;
+    last_v = curr_v;
+  }
+}
+
+int32_t QuantizedBinaryInput::bin_search_encoding(float value)
+{
+  int32_t left = 0, right = quantized_levels - 1;
+
+  while (left < right)
+  {
+    uint32_t middle = (left + right + 1) >> 1;
+
+    if (quantization_thresholds[middle] < value)
+    {
+      left = middle;
+    }
+    else
+    {
+      right = middle - 1;
+    }
+  }
+
+  return left;
+}
+
+void QuantizedBinaryInput::quantize_and_pack(const float *input)
+{
+  Shape shape = {encode_bits, input_bin_size};
+  for (int32_t i = 0; i < input_float_size; ++i)
+  {
+    // encode value of input
+    int32_t encoding = bin_search_encoding(input[i]);
+
+    // pack value to binary buffer
+    int32_t idx = i >> 5;    // divide to 32
+    int32_t offset = i & 31; // mod of 32
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      input_binary[calcOffset(shape, b, idx)] |= ((encoding >> b) & 1) << offset;
+    }
+  }
+}
```

Comment:
This method encodes single vector of input data (single batch) to binary buffer.